### ISSUE TITLE
Effect not showing up in Resolume Arena in OSX 

### ISSUE THREAD

**Issue by DanielArnett** [DanielArnett/360-VJ#13]
I got an email saying that Resolume Arena 7.15 isn't showing the effect even though it's in the correct folder. @xinaesthete when you did the build for OSX do you know what version you tested with? And which did you test with Avenue or Arena?

**Comment by xinaesthete** [comment]
I'll try to check tomorrow. Actually I've got a gig coming up so it'd be good to get back into it a bit. Ping me again to remind me if I don't update you in the next few days.

**Comment by xinaesthete** [comment]
The version I had when I tested earlier was Avenue 7.13.2 (which is the latest version I have a licence for). I've just launched it and it tells me 7.16.0 is available.

I installed 7.16.0 and I'm pretty sure nothing is missing from the plugin list, but the user was probably confused by the plugin name (as am I somewhat). On my system, I have various things including `Reprojection` which I think is the latest build (it has all the parameters for InputProjection / OutputProjection etc. and seems to be working). They were probably looking for `360-VJ`.

I can dig out the project and change the name of the build target if you like.

**Comment by DanielArnett** [comment]
@jorisdejong do you have any advice here? I tried using the latest release plugin, how could it be working on OSX for them but not for myself and someone else.

**Comment by jorisdejong** [comment]
OSX requires everything to be codesigned and notarized. If the person
creating the binary didn't codesign or codesigned incorrectly, the binary
will work on the build computer but nowhere else...

**Comment by DanielArnett** [comment]
Fantastic! Unfortunately my machine is so out of date I can't use XCode anymore so I can't do the build without a pretty significant effort. @xinaesthete do you have any time to recompile with your code signed? [I think a guide like this can show you how to set it up.](https://blog.codemagic.io/how-to-code-sign-publish-ios-apps/)

**Comment by xinaesthete** [comment]
Should be able to manage that @DanielArnett

**Comment by DanielArnett** [comment]
Thanks! The mac I have access to has some very niche software to do its job and I don't want to risk breaking something with a system update if I don't have to.

**Comment by xinaesthete** [comment]
Hi @DanielArnett, just getting around to this now... I've distributed iPad apps a little bit before, so been through some of this which used to be a real pain and can now mostly be done fairly automatically... hopefully the "signing authority" `Authority=Apple Development: Peter Todd (3B4PAZ7888)` is the right one to use (distributing through the store they obviously check lots of stuff in the process - I ran a `codesign` command line tool and it seems to make the right kind of noises).
[Reprojection.bundle.zip](https://github.com/DanielArnett/360-VJ/files/12290862/Reprojection.bundle.zip)

I also made a slight change so the plugin name that appears in Resolume is `360-VJ`, but when I changed the filename manually to match it seemed to stop it showing up (now I think, I suppose that might invalidate the signing).

**Comment by xinaesthete** [comment]
p.s. I'm just trying to figure out the proper way of having this so the project in git isn't associated with my credentials... I could put it into a separate `.xcconfig` file, but that seems like if I `.gitignore` the file itself was ignored it'd cause Xcode to get confused when it was missing on another machine, as there'll still be references to it elsewhere...

I found this

https://stackoverflow.com/questions/68500460/how-do-i-not-commit-the-development-team-lines-in-project-pbxproj-without-desele

which is all very interesting but possibly more than I can bothered with just now... I think the only consequence of adding the `DEVELOPMENT_TEAM=...` will be that anyone else who tried to build the project would get an error message that the certificates associated with that team were unavailable and they'd be prompted to select something else, which'd then lead to a minor conflicting change in git...

**Comment by DanielArnett** [comment]
Weird, I woke up and wondered how this was going, checked my email and saw
this.

Ultimately setting this needs to be in the docs. It'd be cool if we could
'catch' this error and print a nice clean error message. Also this is
something that could be pushed upstream to the original Resolume ffgl code,
or maybe they have a way of addressing it.

**Comment by xinaesthete** [comment]
Almost slipped my mind, I'm sure you'd have reminded me if it had.

Although it's clearly not at all the proper solution, part of me is tempted to push changes with the setting applied, which I think would just cause anyone trying to build to get a fairly descriptive error message pointing them to the setting they'd need to change (and wouldn't leak any secret information). If we have then have something in the doc explaining how to either do an unsigned build for local use, or apply a different signing authority (might want to review what the best place to point to for that would be).

Alternatively, we could leave it so that unless configured there's no error but the resulting build is unsigned, and just mention that in the doc. Maybe that's the better option of the two.

Do you sign the Windows builds? What's the approach there?

FWIW, when Xcode has been setup with certificates for development (which as I say, it's relatively good at managing for you these days, at least compared to how it was), the relevant config options are found by searching for `signing` in build settings:

<img width="960" alt="Screenshot 2023-08-08 at 13 57 48" src="https://github.com/DanielArnett/360-VJ/assets/674834/ef6d1556-fcc8-4b6a-8898-314c5e28781e">

From this list of other available options, it seems possible that I didn't actually have the right kind of certificate for distribution in my keychain, which might mean that someone running this on another machine might still have problems, so it'd be good to get some user feedback: 
<img width="425" alt="Screenshot 2023-08-08 at 14 06 01" src="https://github.com/DanielArnett/360-VJ/assets/674834/69090e34-a3a8-45e4-85fa-5f939aa58d5f">

When I do these builds I still go through the annoying extra step of packaging an 'archive' and then manually extracting the one file I want from it, it'd be nice if that was a bit simpler.

**Comment by DanielArnett** [comment]
Windows builds fine without code signing. Idk if the file in question has
private keys you'd want to protect so I'd check on that. I'd have no
problems with you pushing something like this so long as it doesn't become
a security issue for you. If you could put some steps in the readme for
building on OSX (1. Install version xyz of XCode. 2. Follow this doc on
codesigning, 3. etc) that'd be good enough whether you push your local
files or not, then if anyone complains later I'd point to the README.

**Comment by jorisdejong** [comment]
I happened to be working on another problem on my OSX machine, so I did a quick test. Unfortunately, the zip provided by @xinaesthete does not work for me

**Comment by xinaesthete** [comment]
Thanks @jorisdejong I should hopefully be able to use a different signing authority or something that might fix it. Any more details on what goes wrong? I can probably also arrange so I can better test for myself as well.

**Comment by DanielArnett** [comment]
Same, I'm also not able to run it. I'm making a ticket on the Resolume ffgl
GitHub but keep getting distracted.

**Comment by jorisdejong** [comment]
On compiled binaries, I only have the Resolume log to work with. It tries to load the bundle and then complains that the lib can't be opened, which is a catch all error message
```
[2023-08-08T20:18:25.621+02:00]   INFO:         load: Loading plugin '/Users/hybridair/Documents/Resolume Arena/Extra Effects/Reprojection.bundle/Contents/MacOS/Reprojection'
[2023-08-08T20:18:25.621+02:00]  ERROR:             : Loading plugin failed, lib could not be opened.
```

**Comment by DanielArnett** [comment]
[Okay I pushed this issue upstream.](https://github.com/resolume/ffgl/issues/97)

**Comment by yannicks-png** [comment]
@jorisdejong just asked me to share some of my knowledge. I also had these problems with my plugins. It's been a while so don't be harsh on me if something's not up to date anymore.

In Xcode I logged into my Apple developer account (where I also am a member, which you need to sign and notarize your software). I signed it with my team and with a "Developer ID Application" certificate. Then after it's build you zip the .bundle file and use the notarytool for submitting it to Apple for notarization. That's pretty straight forward. You save the id they give you and they send you an e-mail when the process is finished. With the finished notarization they give you a ticket. You need to use the stapler tool to staple it to your bundle. Then it should work if you distribute that file.

Here's the guide: https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution/customizing_the_notarization_workflow#:~:text=Use%20xcrun%20to%20invoke%20the%20notarytool%20command%20with%20the%20submit%20option%3A

**Comment by DanielArnett** [comment]
@xinaesthete it sounds like the notarytool piece is what you missed.

**Comment by xinaesthete** [comment]
@yannicks-png thanks for that, fingers crossed we'll get it sorted soon.

**Comment by xinaesthete** [comment]
I'm not sure I'll be able to get this done today, which means it'll probably end up being next week.

**Comment by DanielArnett** [comment]
@xinaesthete no worries. That's still way faster than I can update the planetarium's OSX.

**Comment by thedoger82** [comment]
Thanks for answering!!! 
Daniel, do you think that if i share thru anydesk or another remotedesktop software, my screen/computer to you, you would be able to fix it?

**Comment by DanielArnett** [comment]
Hmm that could work; we could build the code locally so it would only work
on your computer. The problem is that Apple also requires devs to be a part
of the Apple Developer Program which costs 99 per year. So even if I helped
you build this, it would only work on your machine.

**Comment by thedoger82** [comment]
I think we can build one in my computer, and then if im not wrong anyone can fix it with this:

xcode-select --install
	sudo codesign --force --deep --sign - /Path/Plugin.bundle
sudo xattr -d -r com.apple.quarantine /Path/Plugin.bundle
	sudo chmod +x /Path/Plugin.bundle

Let me know whats the best platform for you to share screen and when

**Comment by thedoger82** [comment]
Yes I do what I dont have is the sharing soft

**Comment by DanielArnett** [comment]
Send me an email
***@***.***

**Comment by thedoger82** [comment]
Cool, anydesk then, i remember i got banned in teamviewer once hihihi

**Comment by thedoger82** [comment]
this is my anydesk address: 1351432183

**Comment by thedoger82** [comment]
my email is my username at gmail

**Comment by jorisdejong** [comment]
@DanielArnett are you open to hosting just the h and cpp files in this repo?  That way, I can probably make this plugin part of my build pipeline, which has automated signing and notarizing. People that just want the plugins could then download signed binaries for Windows and Mac from the Releases tab. People that want to build from source wouldn't have Xcode and Visual Studio projects anymore, but the code itself would still be open source and if they know what they're doing, they could still add them to the FFGL project themselves and build from there

**Comment by DanielArnett** [comment]
Yeah that'd be good. How about I set this repo to just be 4 files

README.md
Reprojection.cpp
Reprojection.h
Reprojection.hlsl (my shader).

Is that what you're thinking?

I also have some updates I'd like to make here relatively soon (adding
stereoscopic support on another ticket)

**Comment by jorisdejong** [comment]
Yes that's perfect! I'll need to add a cmakelists.txt file to integrate it with my build system, but after that, making new builds with your changes should be a matter of a few clicks

**Comment by DanielArnett** [comment]
The binaries are now up to date and working on macs.
